bionic/code_references.py

```diff
@@ -0,0 +1,214 @@
+            my_cls.call()
+
+
+    On a higher level, for Smart Caching, this leaves some changes that Bionic
+    won't be able to detect. Bionic should be able to detect any changes except
+    when a function returns a class or a module. In those case, Bionic cannot
+    detect what is returned. But it should be able to detect the class / module
+    returned when hashing the function which returns the said result. Bionic will
+    still hash those classes / modules, and detect as many changes as it can in
+    them.
+    """
+    # Top of the stack.
+    tos = None
+    lineno = None
+    refs = []
+
+    def set_tos(t):
+        nonlocal tos
+        if tos is not None:
+            # If the top of stack item already exists, that means we
+            # have gone through all the instructions that use the item,
+            # and it is a reference object.
+            refs.append(tos)
+        tos = t
+
+    # Our goal is to find referenced objects. The problem is that co_names
+    # does not have full qualified names in it. So if you access `foo.bar`,
+    # co_names has `foo` and `bar` in it but it doesn't tell us that the
+    # code reads `bar` of `foo`. We are going over the bytecode to resolve
+    # from which object an attribute is requested.
+    # Read more about bytecode at https://docs.python.org/3/library/dis.html
+
+    for op in dis.get_instructions(code):
+        try:
+            # Sometimes starts_line is None, in which case let's just remember the
+            # previous start_line (if any). This way when there's an exception we at
+            # least can point users somewhat near the line where the error stems from.
+            if op.starts_line is not None:
+                lineno = op.starts_line
+
+            if op.opname in ["LOAD_GLOBAL", "LOAD_NAME"]:
+                if op.argval in context.globals:
+                    set_tos(context.globals[op.argval])
+                else:
+                    # This can happen if the variable does not exist. If we
+                    # cannot find the variable, we return it's name instead.
+                    set_tos(op.argval)
+            elif op.opname in ["LOAD_DEREF", "LOAD_CLOSURE"]:
+                set_tos(context.cells[op.argval])
+            elif op.opname == "IMPORT_NAME":
+                # This instruction only appears if the code object imports a
+                # module using the import statement. If a user is importing
+                # modules inside a function, they probably don't want to import
+                # the module until the function execution time.
+                message = oneline(
+                    f"""
+                Bionic found a code reference in file ${code.co_filename} at line
+                ${lineno} that imports ${op.argval} module when hashing
+                ${code.co_name}. Changes inside the module will likely not be
+                detected. To detect changes inside the module, import it globally
```

Comment:
What if they can't import it globally for some reason? What should they do to avoid this warning?